### PR TITLE
[Cisco Duo] New data stream for Trust Monitor logs

### PR DESCRIPTION
**Pull request by chemamartinez**
## Proposed commit message

Added new data stream `trust_monitor` to collect [Trust Monitor logs](https://duo.com/docs/adminapi#trust-monitor) from Cisco Duo.

The CEL program follows the API specifications set at:
  - [Authentication](https://duo.com/docs/adminapi#authentication)
  - [Response format](https://duo.com/docs/adminapi#response-format)
  - [Parameters and pagination](https://duo.com/docs/adminapi#trust-monitor)

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

Added pipeline and system tests for the new data stream:

```
Run asset tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                                                          │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-0607d4a3-5322-41c1-b8fa-f0d29bcc2757 is loaded │ PASS   │      2.959µs │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-5a0b80af-49ad-42ee-89b7-c89faa927826 is loaded │ PASS   │        250ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-7a135061-78a3-45d9-951b-4b9b665fa729 is loaded │ PASS   │        542ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-b386f94c-0856-4508-ba08-a525a2f3b70f is loaded │ PASS   │        209ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-c3336a66-68ff-4bcd-95ff-fb388793f721 is loaded │ PASS   │        208ns │
│ cisco_duo │                    │ asset     │ dashboard cisco_duo-e91470e5-2ded-4ff1-8bb5-24e06b949c1d is loaded │ PASS   │        500ns │
│ cisco_duo │                    │ asset     │ map cisco_duo-158c0e80-148c-11ec-9386-31989719f9db is loaded       │ PASS   │        209ns │
│ cisco_duo │                    │ asset     │ search cisco_duo-64869c89-8c44-4644-a84a-9815c0fddba0 is loaded    │ PASS   │        208ns │
│ cisco_duo │ admin              │ asset     │ index_template logs-cisco_duo.admin is loaded                      │ PASS   │        375ns │
│ cisco_duo │ admin              │ asset     │ ingest_pipeline logs-cisco_duo.admin-2.1.0 is loaded               │ PASS   │        375ns │
│ cisco_duo │ auth               │ asset     │ index_template logs-cisco_duo.auth is loaded                       │ PASS   │        416ns │
│ cisco_duo │ auth               │ asset     │ ingest_pipeline logs-cisco_duo.auth-2.1.0 is loaded                │ PASS   │        375ns │
│ cisco_duo │ offline_enrollment │ asset     │ index_template logs-cisco_duo.offline_enrollment is loaded         │ PASS   │        458ns │
│ cisco_duo │ offline_enrollment │ asset     │ ingest_pipeline logs-cisco_duo.offline_enrollment-2.1.0 is loaded  │ PASS   │        459ns │
│ cisco_duo │ summary            │ asset     │ index_template logs-cisco_duo.summary is loaded                    │ PASS   │        458ns │
│ cisco_duo │ summary            │ asset     │ ingest_pipeline logs-cisco_duo.summary-2.1.0 is loaded             │ PASS   │        167ns │
│ cisco_duo │ telephony          │ asset     │ index_template logs-cisco_duo.telephony is loaded                  │ PASS   │        459ns │
│ cisco_duo │ telephony          │ asset     │ ingest_pipeline logs-cisco_duo.telephony-2.1.0 is loaded           │ PASS   │        125ns │
│ cisco_duo │ telephony_v2       │ asset     │ index_template logs-cisco_duo.telephony_v2 is loaded               │ PASS   │        459ns │
│ cisco_duo │ telephony_v2       │ asset     │ ingest_pipeline logs-cisco_duo.telephony_v2-2.1.0 is loaded        │ PASS   │        125ns │
│ cisco_duo │ trust_monitor      │ asset     │ index_template logs-cisco_duo.trust_monitor is loaded              │ PASS   │        167ns │
│ cisco_duo │ trust_monitor      │ asset     │ ingest_pipeline logs-cisco_duo.trust_monitor-2.1.0 is loaded       │ PASS   │        125ns │
╰───────────┴────────────────────┴───────────┴────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run pipeline tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                                              │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼────────────────────────────────────────────────────────┼────────┼──────────────┤
│ cisco_duo │ admin              │ pipeline  │ (ingest pipeline warnings test-admin.log)              │ PASS   │ 212.113125ms │
│ cisco_duo │ admin              │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 199.902667ms │
│ cisco_duo │ admin              │ pipeline  │ test-admin.log                                         │ PASS   │ 187.260917ms │
│ cisco_duo │ admin              │ pipeline  │ test-empty.log                                         │ PASS   │  42.235833ms │
│ cisco_duo │ auth               │ pipeline  │ (ingest pipeline warnings test-auth.log)               │ PASS   │ 214.673917ms │
│ cisco_duo │ auth               │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 202.171875ms │
│ cisco_duo │ auth               │ pipeline  │ test-auth.log                                          │ PASS   │ 363.385041ms │
│ cisco_duo │ auth               │ pipeline  │ test-empty.log                                         │ PASS   │  41.348083ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 215.647541ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ (ingest pipeline warnings test-offline-enrollment.log) │ PASS   │  199.06175ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ test-empty.log                                         │ PASS   │  41.397125ms │
│ cisco_duo │ offline_enrollment │ pipeline  │ test-offline-enrollment.log                            │ PASS   │   41.71375ms │
│ cisco_duo │ summary            │ pipeline  │ (ingest pipeline warnings test-summary.log)            │ PASS   │ 194.195792ms │
│ cisco_duo │ summary            │ pipeline  │ test-summary.log                                       │ PASS   │  41.787834ms │
│ cisco_duo │ telephony          │ pipeline  │ (ingest pipeline warnings test-empty.log)              │ PASS   │ 202.682416ms │
│ cisco_duo │ telephony          │ pipeline  │ (ingest pipeline warnings test-telephony.log)          │ PASS   │   190.2915ms │
│ cisco_duo │ telephony          │ pipeline  │ test-empty.log                                         │ PASS   │  37.114459ms │
│ cisco_duo │ telephony          │ pipeline  │ test-telephony.log                                     │ PASS   │   38.82075ms │
│ cisco_duo │ telephony_v2       │ pipeline  │ (ingest pipeline warnings test-telephony-v2.log)       │ PASS   │  215.32675ms │
│ cisco_duo │ telephony_v2       │ pipeline  │ test-telephony-v2.log                                  │ PASS   │  45.509083ms │
│ cisco_duo │ trust_monitor      │ pipeline  │ (ingest pipeline warnings test-trust-monitor.log)      │ PASS   │ 205.206959ms │
│ cisco_duo │ trust_monitor      │ pipeline  │ test-trust-monitor.log                                 │ PASS   │  49.902208ms │
╰───────────┴────────────────────┴───────────┴────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run policy tests for the package
--- Test results for package: cisco_duo - START ---
No test results
--- Test results for package: cisco_duo - END   ---
Done
Run static tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├───────────┼────────────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ cisco_duo │ admin              │ static    │ Verify sample_event.json │ PASS   │  43.220709ms │
│ cisco_duo │ auth               │ static    │ Verify sample_event.json │ PASS   │  58.603541ms │
│ cisco_duo │ offline_enrollment │ static    │ Verify sample_event.json │ PASS   │  40.225625ms │
│ cisco_duo │ summary            │ static    │ Verify sample_event.json │ PASS   │  37.368667ms │
│ cisco_duo │ telephony          │ static    │ Verify sample_event.json │ PASS   │  38.672125ms │
│ cisco_duo │ telephony_v2       │ static    │ Verify sample_event.json │ PASS   │  37.763334ms │
│ cisco_duo │ trust_monitor      │ static    │ Verify sample_event.json │ PASS   │  49.630833ms │
╰───────────┴────────────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_duo - END   ---
Done
Run system tests for the package
--- Test results for package: cisco_duo - START ---
╭───────────┬────────────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE   │ DATA STREAM        │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├───────────┼────────────────────┼───────────┼───────────┼────────┼───────────────┤
│ cisco_duo │ admin              │ system    │ default   │ PASS   │ 32.843713458s │
│ cisco_duo │ auth               │ system    │ default   │ PASS   │ 29.915677625s │
│ cisco_duo │ offline_enrollment │ system    │ default   │ PASS   │ 32.776141542s │
│ cisco_duo │ summary            │ system    │ default   │ PASS   │ 31.659921625s │
│ cisco_duo │ telephony          │ system    │ default   │ PASS   │ 32.663194792s │
│ cisco_duo │ telephony_v2       │ system    │ default   │ PASS   │ 34.731503375s │
│ cisco_duo │ trust_monitor      │ system    │ default   │ PASS   │ 38.751648833s │
╰───────────┴────────────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: cisco_duo - END   ---
Done
```

## Related issues

- Closes https://github.com/elastic/integrations/issues/10961

## Screenshots

**Data stream configuration**

Screenshot 2024-10-04 at 09 48 14

**Dashboard**

dashboard-trust-monitor

### CODE REVIEW

File: packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        - set:
+            field: event.kind
+            value: pipeline_error
+        - append:
+            field: error.message
+            value: |-
+              Processor "{{{ _ingest.on_failure_processor_type }}}" with tag "{{{ _ingest.on_failure_processor_tag }}}" in pipeline "{{{ _ingest.on_failure_pipeline }}}" failed with message "{{{ _ingest.on_failure_message }}}"
+  - drop:
+      description: Drop if no timestamp (invalid json)
+      if: ctx.json?.surfaced_timestamp == null
```
Comment (efd6) [inline packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml:29]: ISTM that the combination of these is that if the first fails, we drop the document. Is that what we want? If we remove the `on_failure` from the `json` processor and possibly add an `error.message` in the case that there is no timestamp (or just `fail`) that would prevent silently allowing this kind of condition to go unnoticed.

File: packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            value: |-
+              Processor "{{{ _ingest.on_failure_processor_type }}}" with tag "{{{ _ingest.on_failure_processor_tag }}}" in pipeline "{{{ _ingest.on_failure_pipeline }}}" failed with message "{{{ _ingest.on_failure_message }}}"
+  - fingerprint:
+      fields:
+        - json.sekey
+        - json.surfaced_timestamp
+      target_field: _id
+      ignore_missing: true
+
+  ## Custom fields
```
Comment (efd6) [inline packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml:46]: Add tags to these processors?

File: packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (chemamartinez) [reply packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml:29]: Good point, I just chose to remove this `drop` processor so any invalid document wouldn't be dropped and the `on_failure` case of the `json` processor would let us know. However, let me know if you think it is better to let the `drop` processor and add a failure condition.

File: packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (chemamartinez) [reply packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml:46]: Yes, I usually add tags to processors that potentially can fail the most, but happy to hear suggestions on this.

File: packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (efd6) [reply packages/cisco_duo/data_stream/trust_monitor/elasticsearch/ingest_pipeline/default.yml]: I think what you have now is fine.

**Review by efd6 — APPROVED** [review]
Mod [this comment](https://github.com/elastic/integrations/pull/11394#discussion_r1801995828).

**Review by efd6 — CHANGES_REQUESTED** [review]
Something about real-world Cisco Duo documents is causing breakage with this code (as it exists in auth). So blocking.

**Comment by chemamartinez** [comment]
> Something about real-world Cisco Duo documents is causing breakage with this code (as it exists in auth). So blocking.

I've checked the case, appreciated that you jumped into it. It is a tricky but relevant inconsistency in the response schema for these endpoints. According to the examples shown in the API docs, the Authentication endpoint is the only one that returns the `next_offset` as an array instead of a string.

**Comment by efd6** [comment]
> I've checked the case, appreciated that you jumped into it. It is a tricky but relevant inconsistency in the response schema for these endpoints.

Indeed! I noticed this.

Thank you for checking.

I think it would be good to use the convention for non-200 body handling in other packages where we just dump the body in the error with the `<METHOD>:` prefix to give something like `GET:{"error…`. The rationale for this approach is that when things have gone wrong, we can not be sure that anything is valid anymore (even in the case that the error-returned body is specified, we find helpful changes — viz the issue noted above — in behaviour that get in the way of being able to correctly do things that give information to the user and support about what went wrong). So rather trying to be helpful and formatting specific fields, which could conceivably fail if the types of the fields change due to API changes, e.g. if `body.message_detail` becomes an array — arguably a reasonable change to make in the API — the error handling would cause a complete failure rather than putting the body in the logs).

File: packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                    "secret_key": state.secret_key,
+                    "limit": state.limit,
+                    "mintime": state.mintime,
+                    "maxtime": state.maxtime,
+                    "date": now.format(time_layout.RFC1123Z),
+                    "want_more": has(body.?response.?metadata.next_offset),
+                    ?"next_offset": (body.?response.?metadata.next_offset.orValue(null) != null) ?
+                                        optional.of(string(body.response.metadata.next_offset))
+                                    :
+                                        optional.none(),
```
Comment (efd6) [inline packages/cisco_duo/data_stream/trust_monitor/agent/stream/cel.yml.hbs:93]: This indentation looks odd. 

Also
```suggestion
                    ?"next_offset": (body.?response.metadata.next_offset.orValue(null) != null) ?
                                        optional.of(string(body.response.metadata.next_offset))
                                    :
                                        optional.none(),
```
(similar change for the line above)

**Review by efd6 — APPROVED** [review]
LGTM after nit addressed.

**Pull request merged**